SubstraFoundation/substra-backend
# Enable blacklisting rotated refresh tokens

Enables the blacklisting of rotated refresh tokens.

Since the rotation was already activated and the blacklist app already installed, the only thing missing was the `BLACKLIST_AFTER_ROTATION` option, now set to True.

The documentation recommends setting up a cron job that cleans up expired blacklisted tokens. I took inspiration from the add-account operator and added an operator that cleans the token once everyday.

Related documentation:
* https://django-rest-framework-simplejwt.readthedocs.io/en/latest/settings.html#rotate-refresh-tokens
* https://django-rest-framework-simplejwt.readthedocs.io/en/latest/blacklist_app.html

# To test this PR

1. patch the backend with the following diff and launch it

```diff
diff --git a/backend/backend/celery.py b/backend/backend/celery.py
index 8a7fc7bc..56d2dca4 100644
--- a/backend/backend/celery.py
+++ b/backend/backend/celery.py
@@ -38,7 +38,7 @@ def setup_periodic_tasks(sender, **kwargs):

     from users.tasks import flush_expired_tokens

-    period = int(os.environ.get('FLUSH_EXPIRED_TOKENS_TASK_PERIOD', 24 * 3600))
+    period = int(os.environ.get('FLUSH_EXPIRED_TOKENS_TASK_PERIOD', 60))
     sender.add_periodic_task(period, flush_expired_tokens.s(), queue='scheduler',
                              name='flush expired tokens')

diff --git a/backend/backend/settings/deps/restframework.py b/backend/backend/settings/deps/restframework.py
index d38155e4..23cc0f1f 100644
--- a/backend/backend/settings/deps/restframework.py
+++ b/backend/backend/settings/deps/restframework.py
@@ -29,8 +29,8 @@ REST_FRAMEWORK = {
 }

 SIMPLE_JWT = {
-    'ACCESS_TOKEN_LIFETIME': timedelta(minutes=int(os.environ.get('ACCESS_TOKEN_LIFETIME', 24*60))),
-    'REFRESH_TOKEN_LIFETIME': timedelta(minutes=int(os.environ.get('REFRESH_TOKEN_LIFETIME', 24*60*7))),
+    'ACCESS_TOKEN_LIFETIME': timedelta(seconds=2),
+    'REFRESH_TOKEN_LIFETIME': timedelta(seconds=10),
     'ROTATE_REFRESH_TOKENS': True,
     'AUTH_HEADER_TYPES': ('JWT',),
     'BLACKLIST_AFTER_ROTATION': True,
``` 
2. launch the frontend, and click around for a minute. You should see multiple calls to `/user/refresh`
3. open a shell in the backend and execute `python manage.py shell`
4. execute the following python script which should return a queryset with multiple tokens in it
```python 
from rest_framework_simplejwt.token_blacklist.models import BlacklistedToken
BlacklistedToken.objects.all()
```
5. then run our flush command, which would result in no blacklisted tokens (`BlacklistedToken.objects.all()` returns an empty queryset)
```python
from users.tasks import flush_expired_tokens
flush_expired_tokens()
BlacklistedToken.objects.all()
```
6. Then click a few more times in the frontend and wait for the periodic task to run. You can then check that there are no more blacklisted tokens.
